// @aws-cdk/aws-logs/lib/log-stream.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cdk = require("@aws-cdk/cdk");
const logs_generated_1 = require("./logs.generated");
/**
 * Define a Log Stream in a Log Group
 */
class LogStream extends cdk.Construct {
    /**
     * Import an existing LogGroup
     */
    static import(scope, id, props) {
        return new ImportedLogStream(scope, id, props);
    }
    constructor(scope, id, props) {
        super(scope, id);
        const resource = new logs_generated_1.CfnLogStream(this, 'Resource', {
            logGroupName: props.logGroup.logGroupName,
            logStreamName: props.logStreamName
        });
        if (props.retainLogStream !== false) {
            resource.options.deletionPolicy = cdk.DeletionPolicy.Retain;
        }
        this.logStreamName = resource.logStreamName;
    }
    /**
     * Export this LogStream
     */
    export() {
        return {
            logStreamName: new cdk.Output(this, 'LogStreamName', { value: this.logStreamName }).makeImportValue().toString()
        };
    }
}
exports.LogStream = LogStream;
/**
 * An imported LogStream
 */
class ImportedLogStream extends cdk.Construct {
    constructor(scope, id, props) {
        super(scope, id);
        this.props = props;
        this.logStreamName = props.logStreamName;
    }
    export() {
        return this.props;
    }
}
