// @aws-cdk/aws-certificatemanager/lib/certificate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cdk_1 = require("@aws-cdk/cdk");
const certificatemanager_generated_1 = require("./certificatemanager.generated");
const util_1 = require("./util");
/**
 * A certificate managed by AWS Certificate Manager
 *
 * IMPORTANT: if you are creating a certificate as part of your stack, the stack
 * will not complete creating until you read and follow the instructions in the
 * email that you will receive.
 *
 * ACM will send validation emails to the following addresses:
 *
 *  admin@domain.com
 *  administrator@domain.com
 *  hostmaster@domain.com
 *  postmaster@domain.com
 *  webmaster@domain.com
 *
 * For every domain that you register.
 */
class Certificate extends cdk_1.Construct {
    /**
     * Import a certificate
     */
    static import(scope, id, props) {
        return new ImportedCertificate(scope, id, props);
    }
    constructor(scope, id, props) {
        super(scope, id);
        const allDomainNames = [props.domainName].concat(props.subjectAlternativeNames || []);
        const cert = new certificatemanager_generated_1.CfnCertificate(this, 'Resource', {
            domainName: props.domainName,
            subjectAlternativeNames: props.subjectAlternativeNames,
            domainValidationOptions: allDomainNames.map(domainValidationOption),
        });
        this.certificateArn = cert.certificateArn;
        /**
         * Return the domain validation options for the given domain
         *
         * Closes over props.
         */
        function domainValidationOption(domainName) {
            const overrideDomain = props.validationDomains && props.validationDomains[domainName];
            return {
                domainName,
                validationDomain: overrideDomain || util_1.apexDomain(domainName)
            };
        }
    }
    /**
     * Export this certificate from the stack
     */
    export() {
        return {
            certificateArn: new cdk_1.Output(this, 'Arn', { value: this.certificateArn }).makeImportValue().toString()
        };
    }
}
exports.Certificate = Certificate;
/**
 * A Certificate that has been imported from another stack
 */
class ImportedCertificate extends cdk_1.Construct {
    constructor(scope, id, props) {
        super(scope, id);
        this.props = props;
        this.certificateArn = props.certificateArn;
    }
    export() {
        return this.props;
    }
}
