// @aws-cdk/aws-ssm/lib/parameter-store-string.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cdk = require("@aws-cdk/cdk");
/**
 * References a public value in AWS Systems Manager Parameter Store
 *
 * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/dynamic-references.html
 */
class ParameterStoreString extends cdk.Construct {
    constructor(scope, id, props) {
        super(scope, id);
        // We use a different inner construct depend on whether we want the latest
        // or a specific version.
        //
        // * Latest - generate a Parameter and reference that.
        // * Specific - use a Dynamic Reference.
        if (props.version === undefined) {
            // Construct/get a singleton parameter under the stack
            const param = new cdk.Parameter(this, 'Parameter', {
                type: 'AWS::SSM::Parameter::Value<String>',
                default: props.parameterName
            });
            this.stringValue = param.stringValue;
        }
        else {
            // Use a dynamic reference
            const dynRef = new cdk.DynamicReference(this, 'Reference', {
                service: cdk.DynamicReferenceService.Ssm,
                referenceKey: `${props.parameterName}:${props.version}`,
            });
            this.stringValue = dynRef.stringValue;
        }
    }
}
exports.ParameterStoreString = ParameterStoreString;
/**
 * References a secret value in AWS Systems Manager Parameter Store
 *
 * It is not possible to retrieve the "latest" value of a secret.
 * Use Secrets Manager if you need that ability.
 *
 * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/dynamic-references.html
 */
class ParameterStoreSecureString extends cdk.DynamicReference {
    constructor(scope, id, props) {
        super(scope, id, {
            service: cdk.DynamicReferenceService.SsmSecure,
            referenceKey: `${props.parameterName}:${props.version}`,
        });
    }
}
exports.ParameterStoreSecureString = ParameterStoreSecureString;
