// @aws-cdk/aws-route53/test/test.zone-delegation-record.js
"use strict";
const assert_1 = require("@aws-cdk/assert");
const cdk_1 = require("@aws-cdk/cdk");
const lib_1 = require("../lib");
class TestApp {
    constructor() {
        this.app = new cdk_1.App();
        const account = '123456789012';
        const region = 'bermuda-triangle';
        this.app.node.setContext(`availability-zones:${account}:${region}`, [`${region}-1a`]);
        this.stack = new cdk_1.Stack(this.app, 'MyStack', { env: { account, region } });
    }
    synthesizeTemplate() {
        return this.app.synthesizeStack(this.stack.name);
    }
}
module.exports = {
    'Zone Delegation records': {
        NS(test) {
            const app = new TestApp();
            const zone = new lib_1.PublicHostedZone(app.stack, 'HostedZone', { zoneName: 'test.public' });
            new lib_1.ZoneDelegationRecord(zone, 'NS', {
                zone,
                delegatedZoneName: 'foo',
                nameServers: ['ns-1777.awsdns-30.co.uk']
            });
            assert_1.expect(app.synthesizeTemplate()).to(assert_1.exactlyMatchTemplate({
                Resources: {
                    HostedZoneDB99F866: {
                        Type: 'AWS::Route53::HostedZone',
                        Properties: {
                            Name: 'test.public.'
                        }
                    },
                    HostedZoneNS1BB87CC3: {
                        Type: 'AWS::Route53::RecordSet',
                        Properties: {
                            HostedZoneId: {
                                Ref: 'HostedZoneDB99F866'
                            },
                            Name: 'foo.test.public.',
                            ResourceRecords: ['ns-1777.awsdns-30.co.uk.'],
                            Type: 'NS',
                            TTL: '172800'
                        }
                    }
                }
            }));
            test.done();
        }
    }
};
