// @aws-cdk/aws-elasticloadbalancingv2/lib/nlb/network-listener.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cdk = require("@aws-cdk/cdk");
const base_listener_1 = require("../shared/base-listener");
const enums_1 = require("../shared/enums");
const network_target_group_1 = require("./network-target-group");
/**
 * Define a Network Listener
 */
class NetworkListener extends base_listener_1.BaseListener {
    /**
     * Import an existing listener
     */
    static import(scope, id, props) {
        return new ImportedNetworkListener(scope, id, props);
    }
    constructor(scope, id, props) {
        super(scope, id, {
            loadBalancerArn: props.loadBalancer.loadBalancerArn,
            protocol: enums_1.Protocol.Tcp,
            port: props.port,
        });
        this.loadBalancer = props.loadBalancer;
        (props.defaultTargetGroups || []).forEach(this._addDefaultTargetGroup.bind(this));
    }
    /**
     * Load balance incoming requests to the given target groups.
     */
    addTargetGroups(_id, ...targetGroups) {
        // New default target(s)
        for (const targetGroup of targetGroups) {
            this._addDefaultTargetGroup(targetGroup);
            targetGroup.registerListener(this);
        }
    }
    /**
     * Load balance incoming requests to the given load balancing targets.
     *
     * This method implicitly creates an ApplicationTargetGroup for the targets
     * involved.
     *
     * @returns The newly created target group
     */
    addTargets(id, props) {
        if (!this.loadBalancer.vpc) {
            // tslint:disable-next-line:max-line-length
            throw new Error('Can only call addTargets() when using a constructed Load Balancer; construct a new TargetGroup and use addTargetGroup');
        }
        const group = new network_target_group_1.NetworkTargetGroup(this, id + 'Group', {
            deregistrationDelaySec: props.deregistrationDelaySec,
            healthCheck: props.healthCheck,
            port: props.port,
            proxyProtocolV2: props.proxyProtocolV2,
            targetGroupName: props.targetGroupName,
            targets: props.targets,
            vpc: this.loadBalancer.vpc,
        });
        this.addTargetGroups(id, group);
        return group;
    }
    /**
     * Export this listener
     */
    export() {
        return {
            listenerArn: new cdk.Output(this, 'ListenerArn', { value: this.listenerArn }).makeImportValue().toString()
        };
    }
}
exports.NetworkListener = NetworkListener;
/**
 * An imported Network Listener
 */
class ImportedNetworkListener extends cdk.Construct {
    constructor(scope, id, props) {
        super(scope, id);
        this.props = props;
        this.listenerArn = props.listenerArn;
    }
    export() {
        return this.props;
    }
}
