// @aws-cdk/cdk/lib/core/tokens/encoding.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const resolve_1 = require("./resolve");
const unresolved_1 = require("./unresolved");
// Encoding Tokens into native types; should not be exported
/**
 * Central place where we keep a mapping from Tokens to their String representation
 *
 * The string representation is used to embed token into strings,
 * and stored to be able to
 *
 * All instances of TokenStringMap share the same storage, so that this process
 * works even when different copies of the library are loaded.
 */
class TokenMap {
    constructor() {
        this.tokenMap = new Map();
    }
    /**
     * Generate a unique string for this Token, returning a key
     *
     * Every call for the same Token will produce a new unique string, no
     * attempt is made to deduplicate. Token objects should cache the
     * value themselves, if required.
     *
     * The token can choose (part of) its own representation string with a
     * hint. This may be used to produce aesthetically pleasing and
     * recognizable token representations for humans.
     */
    registerString(token, representationHint) {
        const key = this.register(token, representationHint);
        return `${BEGIN_STRING_TOKEN_MARKER}${key}${END_TOKEN_MARKER}`;
    }
    /**
     * Generate a unique string for this Token, returning a key
     */
    registerList(token, representationHint) {
        const key = this.register(token, representationHint);
        return [`${BEGIN_LIST_TOKEN_MARKER}${key}${END_TOKEN_MARKER}`];
    }
    /**
     * Returns a `TokenString` for this string.
     */
    createStringTokenString(s) {
        return new TokenString(s, QUOTED_BEGIN_STRING_TOKEN_MARKER, `[${VALID_KEY_CHARS}]+`, QUOTED_END_TOKEN_MARKER);
    }
    /**
     * Returns a `TokenString` for this string.
     */
    createListTokenString(s) {
        return new TokenString(s, QUOTED_BEGIN_LIST_TOKEN_MARKER, `[${VALID_KEY_CHARS}]+`, QUOTED_END_TOKEN_MARKER);
    }
    /**
     * Replace any Token markers in this string with their resolved values
     */
    resolveStringTokens(s, context) {
        const str = this.createStringTokenString(s);
        const fragments = str.split(this.lookupToken.bind(this));
        // require() here to break cyclic dependencies
        const ret = fragments.mapUnresolved(x => resolve_1.resolve(x, context)).join(require('./cfn-concat').cloudFormationConcat);
        if (unresolved_1.unresolved(ret)) {
            return resolve_1.resolve(ret, context);
        }
        return ret;
    }
    resolveListTokens(xs, context) {
        // Must be a singleton list token, because concatenation is not allowed.
        if (xs.length !== 1) {
            throw new Error(`Cannot add elements to list token, got: ${xs}`);
        }
        const str = this.createListTokenString(xs[0]);
        const fragments = str.split(this.lookupToken.bind(this));
        if (fragments.length !== 1) {
            throw new Error(`Cannot concatenate strings in a tokenized string array, got: ${xs[0]}`);
        }
        return fragments.mapUnresolved(x => resolve_1.resolve(x, context)).values[0];
    }
    /**
     * Find a Token by key
     */
    lookupToken(key) {
        const token = this.tokenMap.get(key);
        if (!token) {
            throw new Error(`Unrecognized token key: ${key}`);
        }
        return token;
    }
    register(token, representationHint) {
        const counter = this.tokenMap.size;
        const representation = (representationHint || `TOKEN`).replace(new RegExp(`[^${VALID_KEY_CHARS}]`, 'g'), '.');
        const key = `${representation}.${counter}`;
        this.tokenMap.set(key, token);
        return key;
    }
}
exports.TokenMap = TokenMap;
const BEGIN_STRING_TOKEN_MARKER = '${Token[';
const BEGIN_LIST_TOKEN_MARKER = '#{Token[';
const END_TOKEN_MARKER = ']}';
const QUOTED_BEGIN_STRING_TOKEN_MARKER = regexQuote(BEGIN_STRING_TOKEN_MARKER);
const QUOTED_BEGIN_LIST_TOKEN_MARKER = regexQuote(BEGIN_LIST_TOKEN_MARKER);
const QUOTED_END_TOKEN_MARKER = regexQuote(END_TOKEN_MARKER);
const VALID_KEY_CHARS = 'a-zA-Z0-9:._-';
/**
 * A string with markers in it that can be resolved to external values
 */
class TokenString {
    constructor(str, quotedBeginMarker, idPattern, quotedEndMarker) {
        this.str = str;
        this.pattern = `${quotedBeginMarker}(${idPattern})${quotedEndMarker}`;
    }
    /**
     * Split string on markers, substituting markers with Tokens
     */
    split(lookup) {
        const re = new RegExp(this.pattern, 'g');
        const ret = new TokenizedStringFragments();
        let rest = 0;
        let m = re.exec(this.str);
        while (m) {
            if (m.index > rest) {
                ret.addLiteral(this.str.substring(rest, m.index));
            }
            ret.addUnresolved(lookup(m[1]));
            rest = re.lastIndex;
            m = re.exec(this.str);
        }
        if (rest < this.str.length) {
            ret.addLiteral(this.str.substring(rest));
        }
        return ret;
    }
    /**
     * Indicates if this string includes tokens.
     */
    test() {
        const re = new RegExp(this.pattern, 'g');
        return re.test(this.str);
    }
}
/**
 * Fragments of a string with markers
 */
class TokenizedStringFragments {
    constructor() {
        this.fragments = new Array();
    }
    get length() {
        return this.fragments.length;
    }
    get values() {
        return this.fragments.map(f => f.type === 'unresolved' ? f.token : f.lit);
    }
    addLiteral(lit) {
        this.fragments.push({ type: 'literal', lit });
    }
    addUnresolved(token) {
        this.fragments.push({ type: 'unresolved', token });
    }
    mapUnresolved(fn) {
        const ret = new TokenizedStringFragments();
        for (const f of this.fragments) {
            switch (f.type) {
                case 'literal':
                    ret.addLiteral(f.lit);
                    break;
                case 'unresolved':
                    const mappedToken = fn(f.token);
                    if (unresolved_1.unresolved(mappedToken)) {
                        ret.addUnresolved(mappedToken);
                    }
                    else {
                        ret.addLiteral(mappedToken);
                    }
                    break;
            }
        }
        return ret;
    }
    /**
     * Combine the resolved string fragments using the Tokens to join.
     *
     * Resolves the result.
     */
    join(concat) {
        if (this.fragments.length === 0) {
            return concat(undefined, undefined);
        }
        const values = this.fragments.map(fragmentValue);
        while (values.length > 1) {
            const prefix = values.splice(0, 2);
            values.splice(0, 0, concat(prefix[0], prefix[1]));
        }
        return values[0];
    }
}
/**
 * Resolve the value from a single fragment
 */
function fragmentValue(fragment) {
    return fragment.type === 'literal' ? fragment.lit : fragment.token;
}
/**
 * Quote a string for use in a regex
 */
function regexQuote(s) {
    return s.replace(/[.?*+^$[\]\\(){}|-]/g, "\\$&");
}
const glob = global;
/**
 * Singleton instance of the token string map
 */
exports.TOKEN_MAP = glob.__cdkTokenMap = glob.__cdkTokenMap || new TokenMap();
function isListToken(x) {
    return typeof (x) === 'string' && exports.TOKEN_MAP.createListTokenString(x).test();
}
exports.isListToken = isListToken;
function containsListToken(xs) {
    return xs.some(isListToken);
}
exports.containsListToken = containsListToken;
