// @aws-cdk/cdk/lib/core/tokens/options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Global options for resolve()
 *
 * Because there are many independent calls to resolve(), some losing context,
 * we cannot simply pass through options at each individual call. Instead,
 * we configure global context at the stack synthesis level.
 */
class ResolveConfiguration {
    constructor() {
        this.options = new Array();
    }
    push(options) {
        this.options.push(options);
        return {
            pop: () => {
                if (this.options.length === 0 || this.options[this.options.length - 1] !== options) {
                    throw new Error('ResolveConfiguration push/pop mismatch');
                }
                this.options.pop();
            }
        };
    }
    get collect() {
        for (let i = this.options.length - 1; i >= 0; i--) {
            const ret = this.options[i].collect;
            if (ret !== undefined) {
                return ret;
            }
        }
        return undefined;
    }
}
exports.ResolveConfiguration = ResolveConfiguration;
const glob = global;
/**
 * Singleton instance of resolver options
 */
exports.RESOLVE_OPTIONS = glob.__cdkResolveOptions = glob.__cdkResolveOptions || new ResolveConfiguration();
