// @aws-cdk/aws-iam/test/test.auto-cross-stack-refs.js
"use strict";
const assert_1 = require("@aws-cdk/assert");
const cdk = require("@aws-cdk/cdk");
const iam = require("../lib");
module.exports = {
    'automatic exports are created when attributes are referneced across stacks'(test) {
        // GIVEN
        const stackWithUser = new cdk.Stack();
        const stackWithGroup = new cdk.Stack();
        const user = new iam.User(stackWithUser, 'User');
        const group = new iam.Group(stackWithGroup, 'Group');
        // WHEN
        group.addUser(user);
        //
        // `group.addUser` adds the group to the user resource definition, so we expect
        // that an automatic export will be created for the group and the user's stack
        // to use ImportValue to import it.
        // note that order of "expect"s matters. we first need to synthesize the user's
        // stack so that the cross stack reference will be reported and only then the
        // group's stack. in the real world, App will take care of this.
        //
        // THEN
        assert_1.expect(stackWithUser).toMatch({
            Resources: {
                User00B015A1: {
                    Type: "AWS::IAM::User",
                    Properties: {
                        Groups: [{ "Fn::ImportValue": "ExportsOutputRefGroupC77FDACD8CF7DD5B" }]
                    }
                }
            }
        });
        assert_1.expect(stackWithGroup).toMatch({
            Outputs: {
                ExportsOutputRefGroupC77FDACD8CF7DD5B: {
                    Value: { Ref: "GroupC77FDACD" },
                    Export: { Name: "ExportsOutputRefGroupC77FDACD8CF7DD5B" }
                }
            },
            Resources: {
                GroupC77FDACD: {
                    Type: "AWS::IAM::Group"
                }
            }
        });
        test.done();
    }
};
