// @aws-cdk/aws-cloudwatch/lib/util.statistic.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const metric_1 = require("./metric");
/**
 * Parse a statistic, returning the type of metric that was used (simple or percentile)
 */
function parseStatistic(stat) {
    const lowerStat = stat.toLowerCase();
    // Simple statistics
    const statMap = {
        average: metric_1.Statistic.Average,
        avg: metric_1.Statistic.Average,
        minimum: metric_1.Statistic.Minimum,
        min: metric_1.Statistic.Minimum,
        maximum: metric_1.Statistic.Maximum,
        max: metric_1.Statistic.Maximum,
        samplecount: metric_1.Statistic.SampleCount,
        n: metric_1.Statistic.SampleCount,
        sum: metric_1.Statistic.Sum,
    };
    if (lowerStat in statMap) {
        return {
            type: 'simple',
            statistic: statMap[lowerStat]
        };
    }
    // Percentile statistics
    const re = /^p([\d.]+)$/;
    const m = re.exec(lowerStat);
    if (m) {
        return {
            type: 'percentile',
            percentile: parseFloat(m[1])
        };
    }
    throw new Error(`Not a valid statistic: '${stat}', must be one of Average | Minimum | Maximum | SampleCount | Sum | pNN.NN`);
}
exports.parseStatistic = parseStatistic;
function normalizeStatistic(stat) {
    const parsed = parseStatistic(stat);
    if (parsed.type === 'simple') {
        return parsed.statistic;
    }
    else {
        // Already percentile. Avoid parsing because we might get into
        // floating point rounding issues, return as-is but lowercase the p.
        return stat.toLowerCase();
    }
}
exports.normalizeStatistic = normalizeStatistic;
