// @aws-cdk/applet-js/lib/applet-helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Determine whether this constructorFunction is going to create an object that inherits from Stack
 *
 * We do structural typing.
 */
function isStackConstructor(constructorFn) {
    // Test for a public method that Stack has
    return constructorFn.prototype.findResource !== undefined;
}
exports.isStackConstructor = isStackConstructor;
/**
 * Extract module name from a NPM package specification
 */
function extractModuleName(packageSpec) {
    const m = /^((?:@[a-zA-Z-]+\/)?[a-zA-Z-]+)/i.exec(packageSpec);
    if (!m) {
        throw new Error(`Could not find package name in ${packageSpec}`);
    }
    return m[1];
}
exports.extractModuleName = extractModuleName;
function parseApplet(applet) {
    const m = /^(npm:\/\/)?([a-z0-9_@./-]+)(:[a-z_0-9]+)?$/i.exec(applet);
    if (!m) {
        throw new Error(`"applet" value is "${applet}" but it must be in the form "[npm://]<js-module>[:<applet-class>]".
      If <applet-class> is not specified, "Applet" is the default`);
    }
    if (m[1] === 'npm://') {
        return {
            npmPackage: m[2],
            moduleName: extractModuleName(m[2]),
            className: className(m[3]),
        };
    }
    else {
        return {
            moduleName: m[2],
            className: className(m[3]),
        };
    }
    function className(s) {
        if (s) {
            return s.substr(1);
        }
        return 'Applet';
    }
}
exports.parseApplet = parseApplet;
