// @aws-cdk/aws-sns/lib/topic.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cdk_1 = require("@aws-cdk/cdk");
const sns_generated_1 = require("./sns.generated");
const topic_base_1 = require("./topic-base");
/**
 * A new SNS topic
 */
class Topic extends topic_base_1.TopicBase {
    constructor(scope, id, props = {}) {
        super(scope, id);
        this.autoCreatePolicy = true;
        const resource = new sns_generated_1.CfnTopic(this, 'Resource', {
            displayName: props.displayName,
            topicName: props.topicName
        });
        this.topicArn = resource.ref;
        this.topicName = resource.topicName;
    }
    /**
     * Import a Topic defined elsewhere
     */
    static import(scope, id, props) {
        return new ImportedTopic(scope, id, props);
    }
    /**
     * Export this Topic
     */
    export() {
        return {
            topicArn: new cdk_1.Output(this, 'TopicArn', { value: this.topicArn }).makeImportValue().toString(),
            topicName: new cdk_1.Output(this, 'TopicName', { value: this.topicName }).makeImportValue().toString(),
        };
    }
}
exports.Topic = Topic;
/**
 * An imported topic
 */
class ImportedTopic extends topic_base_1.TopicBase {
    constructor(scope, id, props) {
        super(scope, id);
        this.props = props;
        this.autoCreatePolicy = false;
        this.topicArn = props.topicArn;
        this.topicName = props.topicName;
    }
    export() {
        return this.props;
    }
}
