// @aws-cdk/cdk/lib/aspects/tag-aspect.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const resource_1 = require("../cloudformation/resource");
/**
 * The common functionality for Tag and Remove Tag Aspects
 */
class TagBase {
    constructor(key, props = {}) {
        this.key = key;
        this.props = props;
    }
    visit(construct) {
        if (!resource_1.Resource.isResource(construct)) {
            return;
        }
        const resource = construct;
        if (resource_1.Resource.isTaggable(resource)) {
            this.applyTag(resource);
        }
    }
}
exports.TagBase = TagBase;
/**
 * The Tag Aspect will handle adding a tag to this node and cascading tags to children
 */
class Tag extends TagBase {
    constructor(key, value, props = {}) {
        super(key, props);
        this.props.applyToLaunchedInstances = props.applyToLaunchedInstances !== false;
        this.props.priority = props.priority === undefined ? 0 : props.priority;
        if (value === undefined) {
            throw new Error('Tag must have a value');
        }
        this.value = value;
    }
    applyTag(resource) {
        resource.tags.setTag(this.key, this.value, this.props);
    }
}
exports.Tag = Tag;
/**
 * The RemoveTag Aspect will handle removing tags from this node and children
 */
class RemoveTag extends TagBase {
    constructor(key, props = {}) {
        super(key, props);
        this.props.priority = props.priority === undefined ? 1 : props.priority;
    }
    applyTag(resource) {
        resource.tags.removeTag(this.key, this.props);
        return;
    }
}
exports.RemoveTag = RemoveTag;
