// aws-cdk/lib/archive.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const archiver = require("archiver");
const crypto = require("crypto");
const fs = require("fs-extra");
function zipDirectory(directory, outputFile) {
    return new Promise((ok, fail) => {
        const output = fs.createWriteStream(outputFile);
        const archive = archiver('zip');
        // The below options are needed to support following symlinks when building zip files:
        // -  nodir: This will prevent symlinks themselves from being copied into the zip.
        // - follow: This will follow symlinks and copy the files within.
        const globOptions = {
            dot: true,
            nodir: true,
            follow: true,
            cwd: directory
        };
        archive.glob('**', globOptions);
        archive.pipe(output);
        archive.finalize();
        archive.on('warning', fail);
        archive.on('error', fail);
        output.once('close', () => ok());
    });
}
exports.zipDirectory = zipDirectory;
function md5hash(data) {
    return crypto.createHash('sha256').update(data).digest('hex');
}
exports.md5hash = md5hash;
