// aws-cdk/lib/renames.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const util = require("./util");
/**
 * A class used to maintain a set of rename directives
 */
class Renames {
    constructor(renameTable, defaultRename) {
        this.renameTable = renameTable;
        this.defaultRename = defaultRename;
    }
    /**
     * Check the selected stacks against the renames, to see that they make sense
     *
     * We explicitly don't check that renamed stacks are in the passed set, so
     * that people may use a default rename table  even when only selecting
     * subsets of stacks.
     *
     * We DO check that if there's a default rename (simple syntax) they
     * only selected one stack.
     */
    validateSelectedStacks(stacks) {
        if (this.hasDefaultRename && stacks.length > 1) {
            throw new Error("When selecting multiple stacks, you must use the 'ORIGINALNAME:RENAME' pattern for renames.");
        }
    }
    /**
     * Whether this rename has a single rename that renames any stack
     */
    get hasDefaultRename() {
        return this.defaultRename != null;
    }
    /**
     * Return the target name for a given stack name
     *
     * Returns either the renamed name or the original name.
     */
    finalName(name) {
        const rename = this.lookupRename(name);
        if (rename != null) {
            return rename;
        }
        return name;
    }
    lookupRename(name) {
        if (this.hasDefaultRename) {
            return this.defaultRename;
        }
        return this.renameTable[name];
    }
}
exports.Renames = Renames;
/**
 * Parse a rename expression string and construct a Renames object from it
 *
 * The rename expression looks like:
 *
 *  [OLD:]NEW[,OLD:NEW[,...]]
 *
 * If there is more than one rename, every entry must have an OLD name.
 */
function parseRenames(renameExpr) {
    if (renameExpr == null || renameExpr.trim().length === 0) {
        return new Renames({});
    }
    const clauses = renameExpr.split(',');
    if (clauses.length === 1 && clauses[0].indexOf(':') === -1) {
        return new Renames({}, clauses[0]);
    }
    const table = util.makeObject(clauses
        .map(s => s.trim())
        .filter(s => s.length > 0)
        .map(clause => clause.split(':', 2)));
    return new Renames(table);
}
exports.parseRenames = parseRenames;
