// @aws-cdk/aws-ec2/lib/vpc-network-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cdk = require("@aws-cdk/cdk");
const cxapi = require("@aws-cdk/cx-api");
/**
 * Context provider to discover and import existing VPCs
 */
class VpcNetworkProvider {
    constructor(context, props) {
        const filter = props.tags || {};
        // We give special treatment to some tags
        if (props.vpcId) {
            filter['vpc-id'] = props.vpcId;
        }
        if (props.vpcName) {
            filter['tag:Name'] = props.vpcName;
        }
        if (props.isDefault !== undefined) {
            filter.isDefault = props.isDefault ? 'true' : 'false';
        }
        this.provider = new cdk.ContextProvider(context, cxapi.VPC_PROVIDER, { filter });
    }
    /**
     * Return the VPC import props matching the filter
     */
    get vpcProps() {
        const ret = this.provider.getValue(DUMMY_VPC_PROPS);
        return ret;
    }
}
exports.VpcNetworkProvider = VpcNetworkProvider;
/**
 * There are returned when the provider has not supplied props yet
 *
 * It's only used for testing and on the first run-through.
 */
const DUMMY_VPC_PROPS = {
    availabilityZones: ['dummy-1a', 'dummy-1b'],
    vpcId: 'vpc-12345',
    publicSubnetIds: ['s-12345', 's-67890'],
    privateSubnetIds: ['p-12345', 'p-67890'],
};
