// aws-cdk/test/test.diff.js
"use strict";
const stream_1 = require("stream");
const string_decoder_1 = require("string_decoder");
const stacks_1 = require("../lib/api/cxapp/stacks");
const sdk_1 = require("../lib/api/util/sdk");
const cdk_toolkit_1 = require("../lib/cdk-toolkit");
const settings_1 = require("../lib/settings");
const FIXED_RESULT = {
    version: '1',
    stacks: [
        {
            name: 'A',
            template: { resource: 'A' },
            environment: { name: 'dev', account: '12345', region: 'here' },
            metadata: {},
        },
        {
            name: 'B',
            template: { resource: 'B' },
            environment: { name: 'dev', account: '12345', region: 'here' },
            metadata: {},
            dependsOn: ['A'],
        }
    ]
};
const appStacks = new stacks_1.AppStacks({
    configuration: new settings_1.Configuration(),
    aws: new sdk_1.SDK(),
    synthesizer: async () => FIXED_RESULT,
});
class StringWritable extends stream_1.Writable {
    constructor(options = {}) {
        super(options);
        this._decoder = new string_decoder_1.StringDecoder(options && options.defaultEncoding);
        this.data = '';
    }
    _write(chunk, encoding, callback) {
        if (encoding === 'buffer') {
            chunk = this._decoder.write(chunk);
        }
        this.data += chunk;
        callback();
    }
    _final(callback) {
        this.data += this._decoder.end();
        callback();
    }
}
module.exports = {
    async 'diff can diff multiple stacks'(test) {
        // GIVEN
        const provisioner = {
            async readCurrentTemplate(_stack) {
                return {};
            },
            async deployStack(_options) {
                return { noOp: true, outputs: {}, stackArn: '' };
            }
        };
        const toolkit = new cdk_toolkit_1.CdkToolkit({ appStacks, provisioner });
        const buffer = new StringWritable();
        // WHEN
        const exitCode = await toolkit.diff({
            stackNames: ['B'],
            stream: buffer
        });
        // THEN
        const plainTextOutput = buffer.data.replace(/\x1B\[[0-?]*[ -/]*[@-~]/g, '');
        test.ok(plainTextOutput.indexOf('Stack A') > -1, `Did not contain "Stack A": ${plainTextOutput}`);
        test.ok(plainTextOutput.indexOf('Stack B') > -1, `Did not contain "Stack B": ${plainTextOutput}`);
        test.equals(1, exitCode);
        test.done();
    },
};
