// aws-cdk/test/api/test.stacks.js
"use strict";
const cxapi = require("@aws-cdk/cx-api");
const lib_1 = require("../../lib");
const stacks_1 = require("../../lib/api/cxapp/stacks");
const renames_1 = require("../../lib/renames");
const settings_1 = require("../../lib/settings");
const FIXED_RESULT = {
    version: '1',
    stacks: [
        {
            name: 'withouterrors',
            template: { resource: 'noerrorresource' },
            environment: { name: 'dev', account: '12345', region: 'here' },
            metadata: {},
        },
        {
            name: 'witherrors',
            template: { resource: 'errorresource' },
            environment: { name: 'dev', account: '12345', region: 'here' },
            metadata: {
                '/resource': [
                    {
                        type: cxapi.ERROR_METADATA_KEY,
                        data: 'this is an error',
                        trace: []
                    }
                ]
            }
        }
    ]
};
module.exports = {
    async 'do not throw when selecting stack without errors'(test) {
        // GIVEN
        const stacks = new stacks_1.AppStacks({
            configuration: new settings_1.Configuration(),
            aws: new lib_1.SDK(),
            synthesizer: async () => FIXED_RESULT,
        });
        // WHEN
        const selected = await stacks.selectStacks(['withouterrors'], stacks_1.ExtendedStackSelection.None);
        // THEN
        test.equal(selected[0].template.resource, 'noerrorresource');
        test.done();
    },
    async 'do throw when selecting stack with errors'(test) {
        // GIVEN
        const stacks = new stacks_1.AppStacks({
            configuration: new settings_1.Configuration(),
            aws: new lib_1.SDK(),
            synthesizer: async () => FIXED_RESULT,
        });
        // WHEN
        try {
            await stacks.selectStacks(['witherrors'], stacks_1.ExtendedStackSelection.None);
            test.ok(false, 'Did not get exception');
        }
        catch (e) {
            test.ok(/Found errors/.test(e.toString()), 'Wrong error');
        }
        test.done();
    },
    async 'renames get applied when stacks are selected'(test) {
        // GIVEN
        const stacks = new stacks_1.AppStacks({
            configuration: new settings_1.Configuration(),
            aws: new lib_1.SDK(),
            synthesizer: async () => FIXED_RESULT,
            renames: new renames_1.Renames({ withouterrors: 'withoutbananas' }),
        });
        // WHEN
        const synthed = await stacks.selectStacks(['withouterrors'], stacks_1.ExtendedStackSelection.None);
        // THEN
        test.equal(synthed[0].name, 'withoutbananas');
        test.equal(synthed[0].originalName, 'withouterrors');
        test.done();
    },
};
