// @aws-cdk/aws-s3/lib/pipeline-actions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const codepipeline = require("@aws-cdk/aws-codepipeline-api");
/**
 * Source that is provided by a specific Amazon S3 object.
 */
class PipelineSourceAction extends codepipeline.SourceAction {
    constructor(props) {
        super(Object.assign({}, props, { provider: 'S3', outputArtifactName: props.outputArtifactName || `Artifact_${props.actionName}_${props.bucket.node.uniqueId}`, configuration: {
                S3Bucket: props.bucket.bucketName,
                S3ObjectKey: props.bucketKey,
                PollForSourceChanges: props.pollForSourceChanges,
            } }));
        this.props = props;
    }
    bind(stage, _scope) {
        if (this.props.pollForSourceChanges === false) {
            this.props.bucket.onPutObject(stage.pipeline.node.uniqueId + 'SourceEventRule', stage.pipeline, this.props.bucketKey);
        }
        // pipeline needs permissions to read from the S3 bucket
        this.props.bucket.grantRead(stage.pipeline.role);
    }
}
exports.PipelineSourceAction = PipelineSourceAction;
/**
 * Deploys the sourceArtifact to Amazon S3.
 */
class PipelineDeployAction extends codepipeline.DeployAction {
    constructor(props) {
        super(Object.assign({}, props, { provider: 'S3', artifactBounds: {
                minInputs: 1,
                maxInputs: 1,
                minOutputs: 0,
                maxOutputs: 0,
            }, configuration: {
                BucketName: props.bucket.bucketName,
                Extract: (props.extract === false) ? 'false' : 'true',
                ObjectKey: props.objectKey,
            } }));
        this.bucket = props.bucket;
    }
    bind(stage, _scope) {
        // pipeline needs permissions to write to the S3 bucket
        this.bucket.grantWrite(stage.pipeline.role);
    }
}
exports.PipelineDeployAction = PipelineDeployAction;
