// @aws-cdk/aws-cognito/lib/user-pool-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cdk = require("@aws-cdk/cdk");
const cognito_generated_1 = require("./cognito.generated");
/**
 * Types of authentication flow
 */
var AuthFlow;
(function (AuthFlow) {
    /**
     * Enable flow for server-side or admin authentication (no client app)
     */
    AuthFlow["AdminNoSrp"] = "ADMIN_NO_SRP_AUTH";
    /**
     * Enable custom authentication flow
     */
    AuthFlow["CustomFlowOnly"] = "CUSTOM_AUTH_FLOW_ONLY";
    /**
     * Enable auth using username & password
     */
    AuthFlow["UserPassword"] = "USER_PASSWORD_AUTH";
})(AuthFlow = exports.AuthFlow || (exports.AuthFlow = {}));
/**
 * Define a UserPool App Client
 */
class UserPoolClient extends cdk.Construct {
    constructor(scope, id, props) {
        super(scope, id);
        const userPoolClient = new cognito_generated_1.CfnUserPoolClient(this, 'Resource', {
            clientName: props.clientName,
            generateSecret: props.generateSecret,
            userPoolId: props.userPool.userPoolId,
            explicitAuthFlows: props.enabledAuthFlows
        });
        this.clientId = userPoolClient.userPoolClientId;
    }
}
exports.UserPoolClient = UserPoolClient;
