// @aws-cdk/aws-kinesis/test/test.subscriptiondestination.js
"use strict";
const assert_1 = require("@aws-cdk/assert");
const logs = require("@aws-cdk/aws-logs");
const cdk = require("@aws-cdk/cdk");
const lib_1 = require("../lib");
module.exports = {
    'stream can be subscription destination'(test) {
        // GIVEN
        const stack = new cdk.Stack();
        const stream = new lib_1.Stream(stack, 'MyStream');
        const logGroup = new logs.LogGroup(stack, 'LogGroup');
        // WHEN
        new logs.SubscriptionFilter(stack, 'Subscription', {
            logGroup,
            destination: stream,
            filterPattern: logs.FilterPattern.allEvents()
        });
        // THEN: subscription target is Stream
        assert_1.expect(stack).to(assert_1.haveResource('AWS::Logs::SubscriptionFilter', {
            DestinationArn: { "Fn::GetAtt": ["MyStream5C050E93", "Arn"] },
            RoleArn: { "Fn::GetAtt": ["MyStreamCloudWatchLogsCanPutRecords58498490", "Arn"] },
        }));
        // THEN: we have a role to write to the Lambda
        assert_1.expect(stack).to(assert_1.haveResource('AWS::IAM::Role', {
            AssumeRolePolicyDocument: {
                Version: '2012-10-17',
                Statement: [{
                        Action: "sts:AssumeRole",
                        Effect: 'Allow',
                        Principal: { Service: { "Fn::Join": ["", ["logs.", { Ref: "AWS::Region" }, ".amazonaws.com"]] } }
                    }],
            }
        }));
        assert_1.expect(stack).to(assert_1.haveResource('AWS::IAM::Policy', {
            PolicyDocument: {
                Version: '2012-10-17',
                Statement: [
                    {
                        Action: "kinesis:PutRecord",
                        Effect: "Allow",
                        Resource: { "Fn::GetAtt": ["MyStream5C050E93", "Arn"] }
                    },
                    {
                        Action: "iam:PassRole",
                        Effect: "Allow",
                        Resource: { "Fn::GetAtt": ["MyStreamCloudWatchLogsCanPutRecords58498490", "Arn"] }
                    }
                ],
            }
        }));
        test.done();
    },
    'cross-account stream can be subscription destination with Destination'(test) {
        // GIVEN
        const sourceStack = new cdk.Stack(undefined, undefined, { env: { account: '12345' } });
        const logGroup = new logs.LogGroup(sourceStack, 'LogGroup');
        const destStack = new cdk.Stack(undefined, undefined, { env: { account: '67890' } });
        const stream = new lib_1.Stream(destStack, 'MyStream');
        // WHEN
        new logs.SubscriptionFilter(sourceStack, 'Subscription', {
            logGroup,
            destination: stream,
            filterPattern: logs.FilterPattern.allEvents()
        });
        // THEN: the source stack has a Destination object that the subscription points to
        assert_1.expect(destStack).to(assert_1.haveResource('AWS::Logs::Destination', {
            TargetArn: { "Fn::GetAtt": ["MyStream5C050E93", "Arn"] },
            RoleArn: { "Fn::GetAtt": ["MyStreamCloudWatchLogsCanPutRecords58498490", "Arn"] },
        }));
        test.done();
    }
};
