// @aws-cdk/aws-ssm/test/test.parameter-store-string.js
"use strict";
const assert_1 = require("@aws-cdk/assert");
const cdk = require("@aws-cdk/cdk");
const ssm = require("../lib");
module.exports = {
    'can reference SSMPS string - specific version'(test) {
        // GIVEN
        const stack = new cdk.Stack();
        // WHEN
        const ref = new ssm.ParameterStoreString(stack, 'Ref', {
            parameterName: '/some/key',
            version: 123
        });
        // THEN
        test.equal(ref.node.resolve(ref.stringValue), '{{resolve:ssm:/some/key:123}}');
        test.done();
    },
    'can reference SSMPS string - latest version'(test) {
        // GIVEN
        const stack = new cdk.Stack();
        // WHEN
        const ref = new ssm.ParameterStoreString(stack, 'Ref', {
            parameterName: '/some/key',
        });
        // THEN
        assert_1.expect(stack).toMatch({
            Parameters: {
                RefParameter407AF5C8: {
                    Type: "AWS::SSM::Parameter::Value<String>",
                    Default: "/some/key"
                }
            }
        });
        test.deepEqual(ref.node.resolve(ref.stringValue), { Ref: 'RefParameter407AF5C8' });
        test.done();
    },
    'can reference SSMPS secure string'(test) {
        // GIVEN
        const stack = new cdk.Stack();
        // WHEN
        const ref = new ssm.ParameterStoreSecureString(stack, 'Ref', {
            parameterName: '/some/key',
            version: 123
        });
        // THEN
        test.equal(ref.node.resolve(ref.stringValue), '{{resolve:ssm-secure:/some/key:123}}');
        test.done();
    },
};
