// @aws-cdk/aws-route53/lib/records/zone-delegation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cdk = require("@aws-cdk/cdk");
const route53_generated_1 = require("../route53.generated");
const _util_1 = require("./_util");
/**
 * A record to delegate further lookups to a different set of name servers
 */
class ZoneDelegationRecord extends cdk.Construct {
    constructor(scope, id, props) {
        super(scope, id);
        const ttl = props.ttl === undefined ? 172800 : props.ttl;
        const resourceRecords = cdk.unresolved(props.nameServers)
            ? props.nameServers // Can't map a string-array token!
            : props.nameServers.map(ns => (cdk.unresolved(ns) || ns.endsWith('.')) ? ns : `${ns}.`);
        new route53_generated_1.CfnRecordSet(this, 'Resource', {
            hostedZoneId: props.zone.hostedZoneId,
            name: _util_1.determineFullyQualifiedDomainName(props.delegatedZoneName, props.zone),
            type: 'NS',
            ttl: ttl.toString(),
            comment: props.comment,
            resourceRecords,
        });
    }
}
exports.ZoneDelegationRecord = ZoneDelegationRecord;
