// @aws-cdk/aws-lambda/test/test.subscriptiondestination.js
"use strict";
const assert_1 = require("@aws-cdk/assert");
const logs = require("@aws-cdk/aws-logs");
const cdk = require("@aws-cdk/cdk");
const lambda = require("../lib");
module.exports = {
    'lambda can be used as metric subscription destination'(test) {
        // GIVEN
        const stack = new cdk.Stack();
        const fn = new lambda.Function(stack, 'MyLambda', {
            code: new lambda.InlineCode('foo'),
            handler: 'index.handler',
            runtime: lambda.Runtime.NodeJS610,
        });
        const logGroup = new logs.LogGroup(stack, 'LogGroup');
        // WHEN
        new logs.SubscriptionFilter(stack, 'Subscription', {
            logGroup,
            destination: fn,
            filterPattern: logs.FilterPattern.allEvents()
        });
        // THEN: subscription target is Lambda
        assert_1.expect(stack).to(assert_1.haveResource('AWS::Logs::SubscriptionFilter', {
            DestinationArn: { "Fn::GetAtt": ["MyLambdaCCE802FB", "Arn"] },
        }));
        // THEN: Lambda has permissions to be invoked by CWL
        assert_1.expect(stack).to(assert_1.haveResource('AWS::Lambda::Permission', {
            Action: "lambda:InvokeFunction",
            FunctionName: { Ref: "MyLambdaCCE802FB" },
            Principal: { "Fn::Join": ["", ["logs.", { Ref: "AWS::Region" }, ".amazonaws.com"]] }
        }));
        test.done();
    }
};
